File: Algorithm/Other/leetcode/82. Remove Duplicates from Sorted List II/GPT/Remove_Duplicates_from_Sorted_List_II.py
# # ✅ 解答

# ```python
from typing import Optional


# LeetCodeの定義に合わせたListNodeクラス
class ListNode:
    def __init__(self, val: int = 0, next: Optional["ListNode"] = None) -> None:
        self.val: int = val
        self.next: Optional["ListNode"] = next


class Solution:
    """
    Remove Duplicates from Sorted List II
    - 与えられたソート済み連結リストから、重複要素をすべて削除し、重複していない要素のみを残す
    """

    # ------------------------------
    # 🚀 競技プログラミング向け実装（性能最優先、エラーハンドリング省略）
    # ------------------------------
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        Time Complexity: O(n)  （各ノードを一度だけ走査）
        Space Complexity: O(1) （ポインタ操作のみ）
        """
        dummy = ListNode(0, head)  # ダミーノード（削除処理を簡単にするため）
        prev = dummy  # 最後に確定したユニーク要素の直前を指す
        curr = head

        while curr:
            # 重複があるかどうかを確認
            if curr.next and curr.val == curr.next.val:
                # 重複が終わる位置までスキップ
                dup_val = curr.val
                while curr and curr.val == dup_val:
                    curr = curr.next
                prev.next = curr  # 重複を飛ばす
            else:
                prev = curr
                curr = curr.next

        return dummy.next

    # ------------------------------
    # 🏢 業務開発向け実装（型安全・入力検証あり）
    # ------------------------------
    def deleteDuplicates_safe(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """
        型チェック・入力検証込みの堅牢版
        """
        if head is not None and not isinstance(head, ListNode):
            raise TypeError("head must be a ListNode or None")

        # 実データなしなら即終了
        if head is None or head.next is None:
            return head

        dummy = ListNode(0, head)
        prev = dummy
        curr = head

        while curr:
            if curr.next and curr.val == curr.next.val:
                dup_val = curr.val
                while curr and curr.val == dup_val:
                    curr = curr.next
                prev.next = curr
            else:
                prev = curr
                curr = curr.next

        return dummy.next


# ```

# ---

# # 🔍 解説

# ## 1. アルゴリズム選定理由

# * **入力はソート済み** → 重複要素は必ず連続する。
# * 一度の線形走査（O(n)）で十分。
# * 余分なデータ構造は不要（O(1)空間）。

# ---

# ## 2. 処理の流れ

# 1. ダミーノードを作成（先頭が削除される場合にも対応可能にするため）。
# 2. `curr` が重複開始点なら、同じ値をすべてスキップ。
# 3. `prev.next` を `curr` に繋ぎ直して重複ノードを除外。
# 4. 重複がなければ `prev` を更新。
# 5. 最後に `dummy.next` を返す。

# ---

# ## 3. 競技プログラミング版 vs 業務版

# * **競技プログラミング版 (`deleteDuplicates`)**

#   * 入力検証や例外処理は省略
#   * シンプルで高速
# * **業務版 (`deleteDuplicates_safe`)**

#   * 型チェック・エラーハンドリング付き
#   * 予期せぬ入力に対して堅牢

# ---

File: Algorithm/Other/leetcode/82. Remove Duplicates from Sorted List II/GPT/Remove_Duplicates_from_Sorted_List_II.js
// # 1. 問題の分析

// ## 競技プログラミング視点

// * 入力は昇順ソート済みの単方向リスト。
// * 「重複をすべて削除し、ユニーク要素だけ残す」問題。
// * 制約: ノード数最大 300 → O(n) アルゴリズムで十分。
// * 破壊的更新で追加メモリを極小化できる。

// ## 業務開発視点

// * メンテナンス性のため、`ListNode` クラス定義を JSDoc 付きで明示。
// * `deleteDuplicates` は純粋関数で副作用を避ける。
// * 入力検証を行い、型不整合時は `TypeError`、制約違反時は `RangeError` を投げる。

// ## JavaScript特有の考慮点

// * V8 最適化: 単純な `while`/`for` ループ、クロージャを避ける。
// * Hidden Class の安定化: `ListNode` は constructor 内でフィールドを固定。
// * 配列操作は不要。ノード参照操作のみで GC コストを削減。

// ---

// # 2. アルゴリズムアプローチ比較

// | アプローチ               | 時間計算量 | 空間計算量 | JS実装コスト | 可読性 | 備考              |
// | ------------------- | ----- | ----- | ------- | --- | --------------- |
// | 方法A: 1パスで重複除去       | O(n)  | O(1)  | 低       | 中   | **最適**、破壊的更新    |
// | 方法B: 配列化→Mapで数え→再構築 | O(n)  | O(n)  | 中       | 高   | 実装は簡単だが余計なメモリ使用 |
// | 方法C: 二重ループ          | O(n²) | O(1)  | 低       | 高   | 小規模なら動くが非効率     |

// ---

// # 3. 選択したアルゴリズムと理由

// * **選択: 方法A (O(n), O(1))**
// * 理由: 制約下で最速、追加メモリ不要、V8 フレンドリー。
// * JS特有の最適化:

//   * `while` ループでリスト走査。
//   * 新規配列/Mapを作らずに既存ノードを直接接続。
//   * `dummy` ノードを導入して先頭削除も簡潔に。

// ---

// # 4. コード実装（solution.js）

// ```javascript
'use strict';

/**
 * Definition for singly-linked list.
 * @constructor
 * @param {number} val
 * @param {ListNode|null} next
 */
function ListNode(val, next = null) {
    /** @type {number} */
    this.val = val;
    /** @type {ListNode|null} */
    this.next = next;
}

/**
 * Remove all duplicates from sorted linked list.
 * @param {ListNode|null} head - The head of the sorted linked list
 * @returns {ListNode|null} - The head of the list with only distinct numbers
 * @throws {TypeError} If input is not a ListNode or null
 * @throws {RangeError} If node count > 300 or value outside [-100,100]
 * @complexity Time O(n), Space O(1)
 */
function deleteDuplicates(head) {
    if (head !== null && !(head instanceof ListNode)) {
        throw new TypeError('Input must be a ListNode or null');
    }

    // Count nodes & validate values
    let count = 0;
    for (let p = head; p !== null; p = p.next) {
        count++;
        if (count > 300) throw new RangeError('Node count exceeds 300');
        if (typeof p.val !== 'number' || !Number.isInteger(p.val)) {
            throw new TypeError('Node value must be an integer');
        }
        if (p.val < -100 || p.val > 100) {
            throw new RangeError('Node value out of range [-100,100]');
        }
    }

    const dummy = new ListNode(0, head);
    let prev = dummy;

    while (head !== null) {
        if (head.next !== null && head.val === head.next.val) {
            // Skip all nodes with the same value
            while (head.next !== null && head.val === head.next.val) {
                head = head.next;
            }
            prev.next = head.next; // cut out duplicates
        } else {
            prev = prev.next;
        }
        head = head.next;
    }

    return dummy.next;
}

module.exports = { ListNode, deleteDuplicates };
// ```

// ---

// # 5. 追加メモ（JS最適化チェックリスト）

// * ✅ `for`/`while` ループ利用、クロージャ排除。
// * ✅ `ListNode` 形状固定で Hidden Class 最適化。
// * ✅ 新規配列/Mapを避け GC 負荷削減。
// * ✅ Pure function (入力検証込み)。

// ---
